In [497]:
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
import selenium.webdriver.support.ui as sui
from bs4 import BeautifulSoup as bs
import requests 
import re
import pandas as pd
import time
import numpy as np
import smtplib, ssl
import email.utils
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
import getpass
import base64

In [501]:
def send_email(subject, message, passw=False, reply=False):
    port = 465  # For SSL
    smtp_server = "smtp.gmail.com"
    sender_email = "fistudentvent@gmail.com"
    receiver_email ="fistudentvent@gmail.com"
    message = f"""\
    Subject: {subject}
    

    {message}"""
    msg = MIMEMultipart(message)
    msg['Reply-To'] = sender_email
    context = ssl.create_default_context()
    if passw:
        password = f.decrypt(passw).decode()
    else:
         password = getpass.getpass()
    if reply:
        msg.add_header('reply-to', receiver_email)
        with smtplib.SMTP_SSL(smtp_server, port, context=context) as server:
            server.login(sender_email, password)
            server.sendmail(msg['Reply-To'], receiver_email, message)
    else:
        with smtplib.SMTP_SSL(smtp_server, port, context=context) as server:
            server.login(sender_email, password)
            server.sendmail(sender_email, receiver_email, message)

In [330]:
def get_main_page_link(search_by="class", search_term="ca-h3"):
    usuable_links = []
    links = soup.findAll('div')
    for link in links:
        try:
            if search_term in link[search_by]:
                usuable_links.append(link["i-stars"])
        except:
            continue
    return usuable_links

for link in links:
    try:
        if "ca-h3" in link["class"]:
            print(link["href"])
    except:
        print("No Class")

# Gathering Data From Yelp based of Term Tutoring

In [918]:
client_id = '5onuGByRauHpN8oN0PEgtw' #Your client ID goes here (as a string)
api_key = 'Ystbu1NrD08vSyMfFl5XxExXnikbKXwv-MtJRUHCSQNjTYWTSpsn7SRkLQUoVE9nf1nr_JCD7oD9kCjQIgMGQbxKFF-HwpB0dresyDzXqx-nztbSq4rhrVq0MHJcXHYx'


def get_all_info(search_term):    
    data = []
    term = search_term
    location = 'NYC'
    SEARCH_LIMIT = 50
    sort_by = "review_count"
    offset = 0
    search_count = SEARCH_LIMIT

    url = 'https://api.yelp.com/v3/businesses/search'

    headers = {
            'Authorization': 'Bearer {}'.format(api_key),
        }

    while search_count >= 0:
        url_params = {
                        'term': term.replace(' ', '+'),
                        'location': location.replace(' ', '+'),
                        'sort_by':sort_by,
                        'limit': SEARCH_LIMIT,
                        'offset':offset
                    }
        yelp_api = requests.get(url, headers=headers, params=url_params)
        data.append(yelp_api.json())
        offset += SEARCH_LIMIT
        search_count = data[0]["total"] - offset
    return data
# data4 = get_all_info()

In [919]:
terms = ["test prep", "after school", "learning center"]

In [920]:
term1 = get_all_info(terms[0])

In [675]:
len(data4)

18

In [691]:
data4[16]

{'businesses': [],
 'total': 892,
 'region': {'center': {'longitude': -73.99429321289062,
   'latitude': 40.70544486444615}}}

In [253]:
def clean_yelp_search(data):
    info = []
    for datum in data:
        for d in datum["businesses"]:
            d["categories"] = list(np.array(d["categories"]).flatten())
        info.extend(datum["businesses"])
    return pd.DataFrame(info)

In [677]:
clean_df = clean_yelp_search(data4)

In [678]:
clean_df

alias  \
0                            [manhattan-prep-new-york]   
1                          [180-degrees-lsat-brooklyn]   
2                           [bright-kids-nyc-new-york]   
3                         [cpr123-queens-forest-hills]   
4                            [veritas-prep-new-york-2]   
5                        [premier-aptitude-flushing-2]   
6                            [talent-prep-manhattan-3]   
7                                    [myguru-new-york]   
8                [project-management-academy-new-york]   
9                        [kweller-prep-forest-hills-2]   
10                          [bee-tutored-park-slope-4]   
11                               [learning-lab-queens]   
12                       [the-study-shack-manhattan-2]   
13    [fastrackids-eye-level-learning-center-brooklyn]   
14                          [iklass-tutoring-queens-2]   
15                          [bricks-tutoring-brooklyn]   
16                    [mathnasium-of-bayside-new-york]   
17                              [idea-tutors-new-york]   
18                   [pye-education-center-elmhurst-2]   
19                           [academics-west-new-york]   
20                            [scholarskills-brooklyn]   
21                     [better-edge-tutoring-new-york]   
22                 [the-brooklyn-math-tutors-brooklyn]   
23                  [insight-tutoring-center-brooklyn]   
24                        [first-step-coding-new-york]   
25                               [sk-academy-brooklyn]   
26                       [cambridge-coaching-new-york]   
27                                 [mini-ivy-new-york]   
28                      [antoine-education-brooklyn-2]   
29              [precision-academic-tutoring-new-york]   
..                                                 ...   
706              [san-diego-math-tutoring-san-diego-2]   
707                               [6402-8-av-brooklyn]   
708                       [college-nannies-tutors-rye]   
709  [kumon-math-and-reading-center-of-mamaroneck-m...   
710                      [perfect-score-llc-oceanport]   
711               [cedar-hill-learning-center-wyckoff]   
712             [chyten-educational-services-millburn]   
713            [club-z-in-home-tutoring-staten-island]   
714              [eye-level-learning-center-paramus-2]   
715                            [mathnasium-of-rye-rye]   
716        [smart-brain-international-new-hyde-park-4]   
717                  [all-technology-now-port-chester]   
718  [witchcraft-educational-learning-awareness-sou...   
719         [huntington-learning-center-carle-place-2]   
720  [kumon-math-and-reading-center-of-valley-strea...   
721            [college-nannies-and-tutors-livingston]   
722                [kids-empowerment-academy-new-york]   
723      [desmond-writing-consultancy-group-bayport-2]   
724  [kumon-math-and-reading-center-of-woodbridge-w...   
725     [eye-level-learning-center-of-iselin-iselin-2]   
726             [huntington-learning-center-baldwin-4]   
727                     [keyport-kids-academy-keyport]   
728  [kumon-math-and-reading-center-of-closter-clos...   
729        [tutor-time-of-east-rockaway-east-rockaway]   
730                [john-p-stevens-high-school-edison]   
731                       [aloha-mind-math-great-neck]   
732                   [the-princeton-review-syosset-3]   
733                    [the-princeton-review-edison-3]   
734                      [2bg-driving-school-branford]   
735                               [moon-prep-smyrna-2]   

                                            categories  \
0    [{'alias': 'tutoring', 'title': 'Tutoring Cent...   
1    [{'alias': 'privatetutors', 'title': 'Private ...   
2    [{'alias': 'tutoring', 'title': 'Tutoring Cent...   
3    [{'alias': 'cprclasses', 'title': 'CPR Classes'}]   
4    [{'alias': 'tutoring', 'title': 'Tutoring Cent...   
5    [{'alias': 'testprep', 'title': 'Test Preparat...   
6    [{'alias': 'testprep', 'title': 'Test Pre

In [679]:
clean_df.shape

(736, 16)

### Shape is 736 before initial clean. We will need to add more companies

In [712]:
def collapse_feature(df, feature):
    try:
        feature_df = df[feature].apply(pd.Series)
        col = [f"{feature}_{i}" for i in range(0, len(feature_df.columns))]
        feature_df.columns =  col
        feature_df = pd.concat([df, feature_df], axis = 1)
        feature_df.drop([feature], axis=1, inplace=True)
    except:
        print(f"No feature: {feature}")
        return df
    return feature_df

def collapse_feature2(df, features):
    features_df1 = df.copy()
    for feature in features:
        try:
            feature_df = df[feature].apply(pd.Series)
            col = [f"{feature}_{i}" for i in range(0, len(feature_df.columns))]
            feature_df.columns =  col
            features_df1 = pd.concat([features_df1, feature_df], axis = 1)
            features_df1.drop([feature], axis=1, inplace=True)
        except:
            print(f"No feature: {feature}")
    return features_df1

In [713]:
df = collapse_feature2(clean_df, ["categories", "location"])

In [714]:
df.drop(["location_7", "distance", "image_url", "alias", "coordinates"], axis=1, inplace=True)


In [715]:
df

display_phone                      id  is_closed  \
0    (212) 721-7400  zv-tJy03CCqZA1OZ05gJCQ      False   
1    (212) 888-5728  HX483Djwn-9NyB2umn4LVA      False   
2    (646) 434-1084  6z63xLs8i0a-R1FMrzL-zw      False   
3    (718) 544-4277  7aOB0AyHOrKkltw50E0frg      False   
4    (800) 925-7737  tiJtgi0vAQULwzVmJDankg      False   
5    (718) 353-9868  HTNDX9rR0ilEPIbEfTlZaQ      False   
6    (347) 699-6895  fcwfK5FB1vkiK_TwVpHsZQ      False   
7    (347) 960-4536  P9t9F9K0cjP8aVkxFKDQ3Q      False   
8    (877) 637-0450  2BagYAYDco9i-63fK2-5kQ      False   
9    (800) 631-1757  dUtKsAhCxpm0kIv7_GksvQ      False   
10   (917) 482-4880  1sbaw0K12UfjVgdqdAvisQ      False   
11   (347) 494-4422  yaMsR14k1mo7NQFYyjhuMQ      False   
12   (646) 331-9456  dTdy0qThbmbrWX6jsnYqPw      False   
13   (718) 260-8100  -ua_DFVKjfinHFvdDMEoOQ      False   
14   (718) 779-1700  ZSRmuLBUE3QddVNzDzuH3w      False   
15   (619) 880-0133  P8AyIcFtVkJEJYgAiiLgKQ      False   
16   (718) 747-6284  lxR1eMrbUG2bgcJCmiVnUg      False   
17   (212) 300-5311  MUqMfXYbavKET4fLyff3Ww      False   
18   (718) 565-1846  AFG9JiBO5Jj822pKTvGK3w      False   
19   (212) 580-0047  CQuq_ji7bHc-8xNdw6k1Gw      False   
20   (718) 258-4647  oHTTB3H7m2V2-vye49CLLA      False   
21   (917) 244-9055  mT0IM-XLr5bsb_TYsQeWaQ      False   
22   (718) 552-0300  TzWLel-mLZ64LmxKJA8Z1w      False   
23   (718) 484-2150  Hi8_K91FEyvNzLJk7wHwcg      False   
24   (415) 937-7634  LyWk6muxYauxaGfhb2Niig      False   
25   (646) 416-0060  FwN1gkaiK33bjxcQ0aCmwA      False   
26   (617) 714-5956  4DvHoLcn6LUdV7ahoxbxcw      False   
27   (646) 657-0350  Yge6F2kkmZLEg9tywCrlGA      False   
28   (917) 909-1524  tyEJG6UnKF0KW3zAWJcwHQ      False   
29   (212) 380-7207  ZhctswwNjys0gV1iFDWUJw      False   
..              ...                     ...        ...   
706  (718) 930-9085  zKqODtk_FAqyYGRlbBX6XQ      False   
707  (718) 680-3013  hsu5VmbwZ9sRwNxUoh4mzw      False   
708  (914) 908-4079  LKD6ZA7zl7CqJQ0Q_cSa3g      False   
709  (914) 873-8338  wNBguebS4KNZXCKLyebGwQ      False   
710  (732) 979-8897  Boikuzww4DzmADoD0YYlJQ      False   
711  (201) 447-0640  lK5bGp6ZsPPX7PXI_7jhOQ      False   
712  (973) 564-5220  nANdP_1TehtK03pmwgZZ0A      False   
713  (718) 984-0373  JdD1n51MffFTU1BTD0OWrw      False   
714  (201) 444-2011  -IY6GnlxUeUavQnH1dLWUA      False   
715  (914) 967-6284  6dyGw_q6RO3jJwmxPI898w      False   
716  (877) 272-4647  RN3UWhHzQ6ZHtDFhDQAz7A      False   
717  (914) 481-5626  arZD2uf_6MMqqrIMPP96uA      False   
718  (516) 528-1192  -ev6KbXvfp1xOnaS6G9-lA      False   
719  (516) 747-5600  xxZrPEbWe978Y41iAkq-zg      False   
720  (516) 872-5772  huxlIoeeazVRv_zHBQtTGg      False   
721  (973) 629-3668  WsdS9ZhrdV6GEYMD4hEFUQ      False   
722  (212) 931-8537  oTZj8qRWasmeULtayXaf0g      False   
723  (917) 532-9460  EOxNfCuaTq4m9mn7ZZM4nA      False   
724  (732) 896-7747  CWmBZS0qg7wuHasc7oNl1A      False   
725  (732) 347-6463  dXUX_U6CWXgSKrEi5YKWLw      False   
726  (516) 867-4041  5dcYfZgMvwFHkRh-i3lNNw      False   
727  (732) 497-5500  n0UDqJpM70S8TDuIZxz-mQ      False   
728  (201) 750-3242  XuI-fHHe9az9LyyQ24D1OA      False   
729  (855) 210-7495  9_LoeswtnNfiV9SzBDHAsw      False   
730  (732) 452-2960  Ha9FpCYGweM8pJDT4q9pIw      False   
731  (917) 325-3088  MFqNVST11OLNxqnbcpGYkA      False   
732  (800) 273-8439  zQ6Kn3RD9Ozmig9i_YCs2g      False   
733  (732) 395-6346  PAXZvoKa7UB1LcIouYMXhA      False   
734  (203) 315-0123  L2OXRW0P7D6G5k3BdfCAMg      False   
735  (800) 791-7476  ZbZQSyeYy8PipcDSc7W37Q      False   

                                               name         phone price  \
0                                    Manhattan Prep  +12127217400   NaN   
1                                  180 Degrees LSAT  +12128885728   NaN   
2                                   Bright Kids NYC  +16464341084   NaN   
3                                     CPR123 Queens  +17185444277   NaN   
4                                 

In [529]:
def set_up_categories(df, categories):
    df1 = df.copy()
    count = 0
    for category in categories:
        df_change = df[category].apply(pd.Series)
        df_change.drop([0], axis=1, inplace=True)
        df_change.columns = [f'{col}_{count}' for col in df_change.columns]
        count += 1
        df1 = pd.concat([df1, df_change], axis=1)
    df1.drop(categories, axis=1, inplace=True)
    return df1

In [716]:
df1 = set_up_categories(df, ['categories_0', 'categories_1','categories_2'])

In [717]:
df1

display_phone                      id  is_closed  \
0    (212) 721-7400  zv-tJy03CCqZA1OZ05gJCQ      False   
1    (212) 888-5728  HX483Djwn-9NyB2umn4LVA      False   
2    (646) 434-1084  6z63xLs8i0a-R1FMrzL-zw      False   
3    (718) 544-4277  7aOB0AyHOrKkltw50E0frg      False   
4    (800) 925-7737  tiJtgi0vAQULwzVmJDankg      False   
5    (718) 353-9868  HTNDX9rR0ilEPIbEfTlZaQ      False   
6    (347) 699-6895  fcwfK5FB1vkiK_TwVpHsZQ      False   
7    (347) 960-4536  P9t9F9K0cjP8aVkxFKDQ3Q      False   
8    (877) 637-0450  2BagYAYDco9i-63fK2-5kQ      False   
9    (800) 631-1757  dUtKsAhCxpm0kIv7_GksvQ      False   
10   (917) 482-4880  1sbaw0K12UfjVgdqdAvisQ      False   
11   (347) 494-4422  yaMsR14k1mo7NQFYyjhuMQ      False   
12   (646) 331-9456  dTdy0qThbmbrWX6jsnYqPw      False   
13   (718) 260-8100  -ua_DFVKjfinHFvdDMEoOQ      False   
14   (718) 779-1700  ZSRmuLBUE3QddVNzDzuH3w      False   
15   (619) 880-0133  P8AyIcFtVkJEJYgAiiLgKQ      False   
16   (718) 747-6284  lxR1eMrbUG2bgcJCmiVnUg      False   
17   (212) 300-5311  MUqMfXYbavKET4fLyff3Ww      False   
18   (718) 565-1846  AFG9JiBO5Jj822pKTvGK3w      False   
19   (212) 580-0047  CQuq_ji7bHc-8xNdw6k1Gw      False   
20   (718) 258-4647  oHTTB3H7m2V2-vye49CLLA      False   
21   (917) 244-9055  mT0IM-XLr5bsb_TYsQeWaQ      False   
22   (718) 552-0300  TzWLel-mLZ64LmxKJA8Z1w      False   
23   (718) 484-2150  Hi8_K91FEyvNzLJk7wHwcg      False   
24   (415) 937-7634  LyWk6muxYauxaGfhb2Niig      False   
25   (646) 416-0060  FwN1gkaiK33bjxcQ0aCmwA      False   
26   (617) 714-5956  4DvHoLcn6LUdV7ahoxbxcw      False   
27   (646) 657-0350  Yge6F2kkmZLEg9tywCrlGA      False   
28   (917) 909-1524  tyEJG6UnKF0KW3zAWJcwHQ      False   
29   (212) 380-7207  ZhctswwNjys0gV1iFDWUJw      False   
..              ...                     ...        ...   
706  (718) 930-9085  zKqODtk_FAqyYGRlbBX6XQ      False   
707  (718) 680-3013  hsu5VmbwZ9sRwNxUoh4mzw      False   
708  (914) 908-4079  LKD6ZA7zl7CqJQ0Q_cSa3g      False   
709  (914) 873-8338  wNBguebS4KNZXCKLyebGwQ      False   
710  (732) 979-8897  Boikuzww4DzmADoD0YYlJQ      False   
711  (201) 447-0640  lK5bGp6ZsPPX7PXI_7jhOQ      False   
712  (973) 564-5220  nANdP_1TehtK03pmwgZZ0A      False   
713  (718) 984-0373  JdD1n51MffFTU1BTD0OWrw      False   
714  (201) 444-2011  -IY6GnlxUeUavQnH1dLWUA      False   
715  (914) 967-6284  6dyGw_q6RO3jJwmxPI898w      False   
716  (877) 272-4647  RN3UWhHzQ6ZHtDFhDQAz7A      False   
717  (914) 481-5626  arZD2uf_6MMqqrIMPP96uA      False   
718  (516) 528-1192  -ev6KbXvfp1xOnaS6G9-lA      False   
719  (516) 747-5600  xxZrPEbWe978Y41iAkq-zg      False   
720  (516) 872-5772  huxlIoeeazVRv_zHBQtTGg      False   
721  (973) 629-3668  WsdS9ZhrdV6GEYMD4hEFUQ      False   
722  (212) 931-8537  oTZj8qRWasmeULtayXaf0g      False   
723  (917) 532-9460  EOxNfCuaTq4m9mn7ZZM4nA      False   
724  (732) 896-7747  CWmBZS0qg7wuHasc7oNl1A      False   
725  (732) 347-6463  dXUX_U6CWXgSKrEi5YKWLw      False   
726  (516) 867-4041  5dcYfZgMvwFHkRh-i3lNNw      False   
727  (732) 497-5500  n0UDqJpM70S8TDuIZxz-mQ      False   
728  (201) 750-3242  XuI-fHHe9az9LyyQ24D1OA      False   
729  (855) 210-7495  9_LoeswtnNfiV9SzBDHAsw      False   
730  (732) 452-2960  Ha9FpCYGweM8pJDT4q9pIw      False   
731  (917) 325-3088  MFqNVST11OLNxqnbcpGYkA      False   
732  (800) 273-8439  zQ6Kn3RD9Ozmig9i_YCs2g      False   
733  (732) 395-6346  PAXZvoKa7UB1LcIouYMXhA      False   
734  (203) 315-0123  L2OXRW0P7D6G5k3BdfCAMg      False   
735  (800) 791-7476  ZbZQSyeYy8PipcDSc7W37Q      False   

                                               name         phone price  \
0                                    Manhattan Prep  +12127217400   NaN   
1                                  180 Degrees LSAT  +12128885728   NaN   
2                                   Bright Kids NYC  +16464341084   NaN   
3                                     CPR123 Queens  +17185444277   NaN   
4                                 

Find all the unique titles and visually look at the titles that are tutoring related

In [718]:
arr_0 = df1.title_0.unique()
arr_1 = df1.title_1.unique()
arr_2 = df1.title_2.unique()
arr_comb = np.union1d(arr_0.astype(str), arr_1.astype(str))
arr_comb = np.union1d(arr_1.astype(str), arr_comb)

In [719]:
arr_comb

array(['Active Life', 'Adult Education', 'Art Classes', 'Art Galleries',
       'Art Schools', 'Art Supplies', 'Arts & Crafts', 'Bookstores',
       'CPR Classes', 'Career Counseling', 'Child Care & Day Care',
       'College Counseling', 'Colleges & Universities',
       'Community Service/Non-Profit', 'Cosmetology Schools',
       'Counseling & Mental Health', 'Cultural Center', 'Dance Schools',
       'Dance Studios', 'Day Camps', 'Dental Hygienists',
       'Driving Schools', 'Editorial Services', 'Education',
       'Educational Services', 'Elementary Schools', 'Financial Advising',
       'Fitness & Instruction', 'Framing', 'Graphic Design',
       'Guitar Stores', 'Gyms', 'Hair Salons',
       'IT Services & Computer Repair', 'Kids Activities',
       'Language Schools', 'Life Coach', 'Middle Schools & High Schools',
       'Montessori Schools', 'Musical Instrument Services',
       'Musical Instruments & Teachers', 'Musicians', 'Nanny Services',
       'Parenting Classes', 'Par

In [722]:
# get a list of titles that are related to tutoring
def keywords(arr):
    lst = []
    hot_words = ["Tutor", "Child", "Kids", "Education", "Community", "Cultural", "Art", "Elementary", "Middle", 
                 "High", "Camps", "Preschool", "Clubs", "Prep"]
    for item in arr:
        for word in hot_words:
            if word in item:
                lst.append(item)
                break
    return lst

In [630]:
keywords(arr_comb)

['Adult Education',
 'Art Classes',
 'Art Galleries',
 'Art Schools',
 'Art Supplies',
 'Arts & Crafts',
 'Child Care & Day Care',
 'Community Service/Non-Profit',
 'Cultural Center',
 'Day Camps',
 'Education',
 'Educational Services',
 'Elementary Schools',
 'Kids Activities',
 'Middle Schools & High Schools',
 'Performing Arts',
 'Preschools',
 'Private Tutors',
 'Social Clubs',
 'Special Education',
 'Summer Camps',
 'Test Preparation',
 'Tutoring Centers']

In [723]:
# checks if the title is related to tutoring or not
def title_coding(df, titles):
    df_4 = df.copy()
    count = 0
    keys = keywords(arr_comb)
    for title in titles:
        encoding = []
        for item in df_4[title].astype(str).to_list():
            flag = False
            for t in keys:
                if t in item:
                    encoding.append(1)
                    flag = True
                    break
            if flag == False:
                encoding.append(0)
            else:
                flag = False
        df_4[f"title_{count}_code"] = encoding
        count += 1
    return df_4
df_2 = title_coding(df3, ["title_0", "title_1", "title_2"])
            
            

In [725]:
df_2[["name","title_0_code", "title_0", "title_1_code", "title_1", "title_2_code", "title_2"]]

name  title_0_code  \
0                                    Manhattan Prep             1   
1                                  180 Degrees LSAT             1   
2                                   Bright Kids NYC             1   
3                                     CPR123 Queens             0   
4                                      Veritas Prep             1   
5                                  Premier Aptitude             1   
6                                       Talent Prep             1   
7                                            MyGuru             1   
8                        Project Management Academy             1   
9                                      Kweller Prep             1   
10                                      Bee Tutored             1   
11                                     Learning Lab             1   
12                                  The Study Shack             1   
13          FasTracKids / Eye Level Learning Center             1   
14                                  iKlass Tutoring             1   
15                                  Bricks Tutoring             1   
16                            Mathnasium of Bayside             1   
17                                      Idea Tutors             1   
18                             PYE Education Center             1   
19                                   Academics West             0   
20                                    Scholarskills             1   
21                             Better Edge Tutoring             1   
22                         The Brooklyn Math Tutors             1   
23                          Insight Tutoring Center             1   
24                                First Step Coding             0   
25                                       SK Academy             1   
26                               Cambridge Coaching             1   
27                                         Mini Ivy             1   
28                                Antoine Education             1   
29                      Precision Academic Tutoring             0   
..                                              ...           ...   
706                         San Diego Math Tutoring             1   
707                                       6402 8 Av             0   
708                        College Nannies + Tutors             1   
709     Kumon Math and Reading Center of Mamaroneck             1   
710                              Perfect Score, LLC             1   
711                      Cedar Hill Learning Center             1   
712                     Chyten Educational Services             1   
713                         Club Z In Home Tutoring             1   
714                       Eye Level Learning Center             1   
715                               Mathnasium of Rye             1   
716                       Smart Brain International             1   
717                              All Technology Now             0   
718       Witchcraft Educational Learning Awareness             1   
719                      Huntington Learning Center             1   
720  Kumon Math and Reading Center of Valley Stream             1   
721                        College Nannies & Tutors             1   
722                        Kids Empowerment Academy             0   
723               Desmond Writing Consultancy Group             0   
724     Kumon Math and Reading Center of Woodbridge             1   
725             Eye Level Learning Center of Iselin             1   
726                      Huntington Learning Center             1   
727                            Keyport Kids Academy             1   
728        Kumon Math and Reading Center of Closter             1   
729                     Tutor Time of East Rockaway             1   
730                      John P Stevens High School             1   
731                                 ALOHA Mind Math             1   
732                            The Princeton 

In [728]:
df_2["counts"] = df_2["title_0_code"] + df_2["title_1_code"] + df_2["title_2_code"]

In [729]:
df_3 = df_2[df_2.counts.values > 0]

In [730]:
df_3.shape

(657, 27)

In [731]:
df_2.shape

(736, 27)

We reduced the our data by almost 100 entries

### Getting Review Data From Yelp

In [732]:
def get_specific_data(driver, class_name, att="title", get_info="class"):
    elems = driver.find_elements_by_class_name(class_name)
    final = []
    if get_info == "att":
        for elem in elems:
            info = elem.get_attribute(att)
            final.append(info)
    else:
        for elem in elems:
            info = elem.text
            final.append(info)
    return final

In [754]:
def get_star_ratings(driver):
    page = driver.page_source
    soup = BeautifulSoup(page)
    meta = soup.findAll("meta")
    stars = []
    for item in str(meta).split("<meta content="):
        if "ratingValue" in item:
            try:
                stars.append(float(item.strip(' itemprop="ratingValue"/>, ')))
            except:
                continue
    return stars

In [762]:
def get_certain_data(driver, subj):
    page = driver.page_source
    soup = BeautifulSoup(page)
    meta = soup.findAll("meta")
    data = []
    for item in str(meta).split("<meta content="):
        if subj in item:
            try:
                data.append(item.strip(f' itemprop="{subj}"/>, '))
            except:
                continue
    return data

In [748]:
def get_user_name(driver):
    page = driver.page_source
    soup = BeautifulSoup(page)
    meta = soup.findAll("meta")
    names = []
    for item in str(meta).split("<meta content="):
        if "author" in item:
            try:
                names.append(item.strip(' itemprop="author"/>, '))
            except:
                continue
    return names

In [740]:
meta

[<meta content="text/html; charset=utf-8" http-equiv="Content-Type"/>,
 <meta content="en" http-equiv="Content-Language"/>,
 <meta content="101 reviews of Manhattan Prep &quot;I had an absolutely amazing experience in Victoria Crutchfield's GRE class, and I'm certain my fellow classmates did as well. I just finished my last class with Victoria and feel confident in my ability attack the…" name="description"/>,
 <meta content="Yelp,recommendation,San Francisco, bay area, local,business,review,friend,restaurant,dentist,doctor,salon,spa,shopping,store,share,community,massage,sushi,pizza,nails,New York,Los Angeles" name="keywords"/>,
 <meta content="width=1020" name="viewport"/>,
 <meta content="Yelp" name="application-name"/>,
 <meta content="https://s3-media3.fl.yelpcdn.com/assets/2/www/img/b8f6d9d556d5/ico/win8-tile.png" name="msapplication-TileImage"/>,
 <meta content="#c41200" name="msapplication-TileColor"/>,
 <meta content="https://www.yelp.com/" name="msapplication-starturl"/>,
 <m

In [865]:
driver = webdriver.Chrome("/Users/flatironschool/Desktop/class-labs/Mod-4/Event-Planner-Project/chromedriver")
driver.get(df.url[0])
m = get_specific_data(driver, "user-passport-info") + get_specific_data(driver, "user-passport-stats")
driver.close()

In [866]:
m

['Chris M.\nManhattan, NY',
 'Peter S.\nManhattan, New York, NY',
 'Sean H.\nSan Francisco, CA',
 'Abby L.\nDurham, NC',
 'Lillian T.\nSan Francisco, CA',
 'Jenn B.\nBrooklyn, NY',
 'Jeff G.\nNew York, NY',
 'Li C.\nNovi, MI',
 'Rose B.\nManhattan, NY',
 'Robert B.\nManhattan, NY',
 'Zoe G.\nWestport, CT',
 'Polina K.\nShort Hills, Millburn, NJ',
 'Philip D.\nWashington, DC',
 'Michael P.\nBaltimore, MD',
 'Jesse P.\nMendota Heights, MN',
 'Natasha G.\nNew York, NY',
 'Joseph L.\nCoronado, CA',
 'Jeremy M.\nWest Village, Manhattan, NY',
 'Julia C.\nCincinnati, OH',
 'Chad M.\nPuyallup, WA',
 '4 friends\n18 reviews\n58 photos',
 '0 friends\n1 review',
 '215 friends\n5 reviews',
 '0 friends\n4 reviews',
 '325 friends\n235 reviews\n1314 photos\nElite ’19',
 '0 friends\n1 review',
 '0 friends\n3 reviews',
 '0 friends\n2 reviews',
 '1 friend\n45 reviews\n8 photos',
 '0 friends\n6 reviews',
 '126 friends\n1 review',
 '0 friends\n1 review',
 '1 friend\n13 reviews',
 '0 friends\n3 reviews',
 '

In [752]:
me

['Chris M.',
 'Peter S.',
 'Sean H.',
 'Abby L.',
 'Lillian T.',
 'Jenn B.',
 'Jeff G.',
 'Li C.',
 'Rose B.',
 'Robert B.',
 'Zoe G.',
 'Polina K.',
 'Philip D.',
 'Michael P.',
 'Jesse P.',
 'Natasha G.',
 'Joseph L.',
 'Jeremy M.',
 'Julia C.',
 'Chad M.']

In [753]:
user_review

["2/23/2019\nI had an absolutely amazing experience in Victoria Crutchfield's GRE class, and I'm certain my fellow classmates did as well. I just finished my last class with Victoria and feel confident in my ability attack the GRE. More than that, I went from dreading math to truly enjoying it thanks to her enthusiasm and tutelage.\n\nVictoria has every quality of a truly great teacher you can possibly imagine and then some - she really cares about where you're at and how to get you to where you want to be in terms of the material, strategies, and score. I started the class a decade removed from GRE level math and now fully fully confident in my ability to learn any concept and apply the strategies we learned in class to the test. She is incredibly mindful of the class's level of comfort with the material as a whole and always checks in to make sure everyone is understanding explanations, concepts, and strategies.\n\nHighly recommended for anyone at any level. Of comfort with the mater

In [867]:
def get_reviews(driver): 
#     user_name = get_certain_data(driver, "author")
    user_info = get_specific_data(driver, "user-passport-info")
    user_passport = get_specific_data(driver, "user-passport-stats")
    review_date = get_certain_data(driver, "datePublished")
    user_review = get_specific_data(driver, "review-content")
    user_star = np.array(get_certain_data(driver, "ratingValue")[1:]).astype(float)
#     review_dict = {"UserName": user_name, "Review" : user_review,  "Star" : user_star}
    return user_info, user_passport, review_date, user_review, user_star

In [868]:
def clean_text(lst, regular=False):
    if regular:
        text_lst = [(" ").join(item[item.index("\n")+1:].split("\n")) for item in lst]
    else:
        text_lst = [item.split("\n") for item in lst]
    return text_lst

In [871]:
def get_all_reviews(df):
    driver = webdriver.Chrome("/Users/flatironschool/Desktop/class-labs/Mod-4/Event-Planner-Project/chromedriver")
    user_names = []
    user_stats = []
    user_dates = []
    user_reviews = []
    user_stars = []
    user_info = []
    runs = 0
    for link in df.url:
        count = 0
        runs += 1
        url = link[:link.index("?")]
        driver.get(url) 
        pages = int(get_specific_data(driver, "page-of-pages")[0][-1])
        while count <= pages*20:
            driver.get(url+f"?start={count}") 
            uname, upass, ud, urevs, us = get_reviews(driver)
            ur = clean_text(urevs, regular=True)
            un = clean_text(uname)
            up = clean_text(upass)
            user_stats.extend(up)
            user_names.extend(un)
            user_dates.extend(ud)
            user_reviews.extend(ur)
            user_stars.extend(us)
            count += 20
        if runs % 100 == 0:
            print("running!")
        break  
        time.sleep(1)
    driver.close()
    return user_names, user_stats, user_dates, user_reviews, user_stars

In [849]:
(" ").join(["1","2"])

'1 2'

In [436]:
url = df2.url[0]

In [445]:
url[:url.index("?")]

'https://www.yelp.com/biz/manhattan-prep-new-york'

In [872]:
u_name, u_stats, u_date, u_rev, u_stars = get_all_reviews(df_3)

In [875]:
u_rev

["I had an absolutely amazing experience in Victoria Crutchfield's GRE class, and I'm certain my fellow classmates did as well. I just finished my last class with Victoria and feel confident in my ability attack the GRE. More than that, I went from dreading math to truly enjoying it thanks to her enthusiasm and tutelage.  Victoria has every quality of a truly great teacher you can possibly imagine and then some - she really cares about where you're at and how to get you to where you want to be in terms of the material, strategies, and score. I started the class a decade removed from GRE level math and now fully fully confident in my ability to learn any concept and apply the strategies we learned in class to the test. She is incredibly mindful of the class's level of comfort with the material as a whole and always checks in to make sure everyone is understanding explanations, concepts, and strategies.  Highly recommended for anyone at any level. Of comfort with the material.",
 'Elaine

In [816]:
len("2018-06-29")

10

In [914]:
cols = ["user_name", "user_location","date", "review", "star_rating"]
df_index = [name[0] for name in u_name]
user_df = pd.DataFrame([[u_name[i][0], u_name[i][1],u_date[i][:10], u_rev[i], u_stars[i]] for i in range(len(u_date))], 
             columns=cols)

In [915]:
pd.concat([user_df, 

user_name                user_location        date  \
0       Chris M.                Manhattan, NY  2019-02-23   
1       Peter S.      Manhattan, New York, NY  2019-04-22   
2        Sean H.            San Francisco, CA  2019-04-03   
3        Abby L.                   Durham, NC  2018-12-29   
4     Lillian T.            San Francisco, CA  2017-12-04   
5        Jenn B.                 Brooklyn, NY  2018-06-20   
6        Jeff G.                 New York, NY  2018-07-17   
7          Li C.                     Novi, MI  2019-01-04   
8        Rose B.                Manhattan, NY  2018-07-31   
9      Robert B.                Manhattan, NY  2019-01-04   
10        Zoe G.                 Westport, CT  2018-12-14   
11     Polina K.    Short Hills, Millburn, NJ  2018-08-13   
12     Philip D.               Washington, DC  2018-07-01   
13    Michael P.                Baltimore, MD  2018-06-28   
14      Jesse P.          Mendota Heights, MN  2018-06-28   
15    Natasha G.                 New York, NY  2018-04-10   
16     Joseph L.                 Coronado, CA  2018-03-20   
17     Jeremy M.  West Village, Manhattan, NY  2018-08-20   
18      Julia C.               Cincinnati, OH  2018-06-29   
19       Chad M.                 Puyallup, WA  2018-06-29   
20          S J.                 New York, NY  2018-06-01   
21          G F.                Manhattan, NY  2018-04-19   
22      Emily A.                 New York, NY  2018-05-20   
23       Jenn A.                Manhattan, NY  2018-01-23   
24      Lucas P.                 New York, NY  2016-09-15   
25    Maurice T.                 Secaucus, NJ  2016-03-05   
26      Darya S.                Manhattan, NY  2015-03-12   
27       Kyle H.                Manhattan, NY  2016-07-27   
28        Jim C.               Providence, RI  2016-07-20   
29    Mashary A.              Los Angeles, CA  2017-08-21   
..           ...                          ...         ...   
71         Jo W.                Manhattan, NY  2013-04-02   
72      Elisa K.                    Miami, FL  2010-05-17   
73       Isha D.            San Francisco, CA  2012-03-14   
74      savvy s.                 New York, NY  2010-01-28   
75    Michael B.                 New York, NY  2010-03-09   
76     Jayson L.                 New York, NY  2006-10-05   
77      April B.                Manhattan, NY  2013-03-13   
78          h k.                Manhattan, NY  2011-07-01   
79          J C.                Manhattan, NY  2011-03-24   
80       Amin M.                 Brooklyn, NY  2010-12-13   
81     Mubeen M.                Rockville, MD  2011-03-08   
82     Itamar B.                Manhattan, NY  2011-09-07   
83       Eran G.                 Brooklyn, NY  2010-03-18   
84       Dana F.                   Queens, NY  2009-11-24   
85        Amy L.              Los Angeles, CA  2012-11-08   
86    Gregory M.              Jersey City, NJ  2010-01-11   
87      Brent C.             Murfreesboro, TN  2010-08-03   
88    Anthony L.                 New York, NY  2009-02-19   
89      Billy L.            Fresh Meadows, NY  2010-04-11   
90       Sara G.                 New York, NY  2010-01-26   
91      Mitch H.            San Francisco, CA  2010-06-23   
92          C L.               Sacramento, CA  2011-07-08   
93          T G.            San Francisco, CA  2009-04-16   
94        Amy E.                 New York, NY  2010-01-31   
95          G K.                   Queens, NY  2009-08-02   
96    Allison H.              Jersey City, NJ  2010-01-13   
97   Ming Min H.                   Boston, MA  2010-02-18   
98      James S.               Washington, DC  2011-01-31   
99     Daniel D.                San Diego, CA  2011-01-13   
100   Michael N.                 New York, NY  2009-03-02   

                                                review  star_rating  
0    I had an absolutely amazing experience in Vict...          5.0  
1    Elaine Loh at Manhattan Prep was a terrific te...          5.0  
2    The Manha

In [917]:
stats_col_name = ["friend_count", "review_count", "photo_count", "elite_status"]
user_df = user_df.join(pd.DataFrame(u_stats, columns=stats_col_name))

In [904]:
user_df

user_location        date  \
Chris M.                   Manhattan, NY  2019-02-23   
Peter S.         Manhattan, New York, NY  2019-04-22   
Sean H.                San Francisco, CA  2019-04-03   
Abby L.                       Durham, NC  2018-12-29   
Lillian T.             San Francisco, CA  2017-12-04   
Jenn B.                     Brooklyn, NY  2018-06-20   
Jeff G.                     New York, NY  2018-07-17   
Li C.                           Novi, MI  2019-01-04   
Rose B.                    Manhattan, NY  2018-07-31   
Robert B.                  Manhattan, NY  2019-01-04   
Zoe G.                      Westport, CT  2018-12-14   
Polina K.      Short Hills, Millburn, NJ  2018-08-13   
Philip D.                 Washington, DC  2018-07-01   
Michael P.                 Baltimore, MD  2018-06-28   
Jesse P.             Mendota Heights, MN  2018-06-28   
Natasha G.                  New York, NY  2018-04-10   
Joseph L.                   Coronado, CA  2018-03-20   
Jeremy M.    West Village, Manhattan, NY  2018-08-20   
Julia C.                  Cincinnati, OH  2018-06-29   
Chad M.                     Puyallup, WA  2018-06-29   
S J.                        New York, NY  2018-06-01   
G F.                       Manhattan, NY  2018-04-19   
Emily A.                    New York, NY  2018-05-20   
Jenn A.                    Manhattan, NY  2018-01-23   
Lucas P.                    New York, NY  2016-09-15   
Maurice T.                  Secaucus, NJ  2016-03-05   
Darya S.                   Manhattan, NY  2015-03-12   
Kyle H.                    Manhattan, NY  2016-07-27   
Jim C.                    Providence, RI  2016-07-20   
Mashary A.               Los Angeles, CA  2017-08-21   
...                                  ...         ...   
Jo W.                      Manhattan, NY  2013-04-02   
Elisa K.                       Miami, FL  2010-05-17   
Isha D.                San Francisco, CA  2012-03-14   
savvy s.                    New York, NY  2010-01-28   
Michael B.                  New York, NY  2010-03-09   
Jayson L.                   New York, NY  2006-10-05   
April B.                   Manhattan, NY  2013-03-13   
h k.                       Manhattan, NY  2011-07-01   
J C.                       Manhattan, NY  2011-03-24   
Amin M.                     Brooklyn, NY  2010-12-13   
Mubeen M.                  Rockville, MD  2011-03-08   
Itamar B.                  Manhattan, NY  2011-09-07   
Eran G.                     Brooklyn, NY  2010-03-18   
Dana F.                       Queens, NY  2009-11-24   
Amy L.                   Los Angeles, CA  2012-11-08   
Gregory M.               Jersey City, NJ  2010-01-11   
Brent C.                Murfreesboro, TN  2010-08-03   
Anthony L.                  New York, NY  2009-02-19   
Billy L.               Fresh Meadows, NY  2010-04-11   
Sara G.                     New York, NY  2010-01-26   
Mitch H.               San Francisco, CA  2010-06-23   
C L.                      Sacramento, CA  2011-07-08   
T G.                   San Francisco, CA  2009-04-16   
Amy E.                      New York, NY  2010-01-31   
G K.                          Queens, NY  2009-08-02   
Allison H.               Jersey City, NJ  2010-01-13   
Ming Min H.                   Boston, MA  2010-02-18   
James S.                  Washington, DC  2011-01-31   
Daniel D.                  San Diego, CA  2011-01-13   
Michael N.                  New York, NY  2009-03-02   

                                                        review  star_rating  
Chris M.     I had an absolutely amazing experience in Vict...          5.0  
Peter S.     Elaine Loh at Manhattan Prep was a terrific te...          5.0  
Sean H.      The Manhattan Prep course was extremely useful...          5.0  
Abby L.      I took both the in-person course and the Advan...          5.0  
Lillian T.   2 check-ins If I could give more stars, I woul...          5.0  
Jenn B.      On my first GMAT practice test, I scored a 600...          4.0  
Jeff G. 

In [383]:
count = 0
for star in u_stars:
    if star != None:
        count += 1
count

38

In [806]:
meta

[<meta content="text/html; charset=utf-8" http-equiv="Content-Type"/>,
 <meta content="en" http-equiv="Content-Language"/>,
 <meta content="101 reviews of Manhattan Prep &quot;I had an absolutely amazing experience in Victoria Crutchfield's GRE class, and I'm certain my fellow classmates did as well. I just finished my last class with Victoria and feel confident in my ability attack the…" name="description"/>,
 <meta content="Yelp,recommendation,San Francisco, bay area, local,business,review,friend,restaurant,dentist,doctor,salon,spa,shopping,store,share,community,massage,sushi,pizza,nails,New York,Los Angeles" name="keywords"/>,
 <meta content="width=1020" name="viewport"/>,
 <meta content="Yelp" name="application-name"/>,
 <meta content="https://s3-media3.fl.yelpcdn.com/assets/2/www/img/b8f6d9d556d5/ico/win8-tile.png" name="msapplication-TileImage"/>,
 <meta content="#c41200" name="msapplication-TileColor"/>,
 <meta content="https://www.yelp.com/" name="msapplication-starturl"/>,
 <m

In [807]:
import sqlite3
connection = sqlite3.connect('reviews.db')
cursor = connection.cursor()

In [808]:
cursor.execute('''
    CREATE table reviews (
        id INTEGER PRIMARY KEY,
        user_name TEXT,
        date TEXT,
        reviews TEXT,
        stars INTEGER
    );''')

In [809]:
cursor.close()

In [788]:
driver = webdriver.Chrome("/Users/flatironschool/Desktop/class-labs/Mod-4/Event-Planner-Project/chromedriver")
driver.get(df.url[0])
page = driver.page_source
p,w,e,r,t = get_reviews(driver)
driver.close()

In [791]:
int(t[0][-1])

6

In [140]:
data[0]["businesses"][0]

{'id': 'HX483Djwn-9NyB2umn4LVA',
 'alias': '180-degrees-lsat-brooklyn',
 'name': '180 Degrees LSAT',
 'image_url': 'https://s3-media2.fl.yelpcdn.com/bphoto/kPNjKkpGQRhWVXUiXBC7SQ/o.jpg',
 'is_closed': False,
 'url': 'https://www.yelp.com/biz/180-degrees-lsat-brooklyn?adjust_creative=5onuGByRauHpN8oN0PEgtw&utm_campaign=yelp_api_v3&utm_medium=api_v3_business_search&utm_source=5onuGByRauHpN8oN0PEgtw',
 'review_count': 47,
 'categories': [{'alias': 'privatetutors', 'title': 'Private Tutors'},
  {'alias': 'testprep', 'title': 'Test Preparation'}],
 'rating': 5.0,
 'coordinates': {'latitude': 40.7168922, 'longitude': -73.9560446},
 'transactions': [],
 'location': {'address1': '205 North 7th',
  'address2': None,
  'address3': '',
  'city': 'Brooklyn',
  'zip_code': '11211',
  'country': 'US',
  'state': 'NY',
  'display_address': ['205 North 7th', 'Brooklyn, NY 11211']},
 'phone': '+12128885728',
 'display_phone': '(212) 888-5728',
 'distance': 3466.0323321358856}

[]

In [ ]:
class WebScraper(object):
    
    def __init__(self, url):
        html_page = requests.get(url) #Make a get request to retrieve the page
        self.soup = BeautifulSoup(html_page.content, 'html.parser') 
    
    def get_links(cls):
        clsa